File: export_player_template/runtime_achievements.js
(function attachRuntimeAchievementTools(global) {
  const CUSTOM_ACHIEVEMENT_PREFIX = "custom:";
  const MAX_ACHIEVEMENT_ID_LENGTH = 64;

  function cleanText(value, fallback = "", maxLength = 240) {
    const text = String(value ?? "").trim();
    return (text || String(fallback ?? "").trim()).slice(0, maxLength);
  }

  function getSafeAchievementAuthorId(value, fallback = "achievement") {
    const normalize = (candidate) =>
      String(candidate ?? "")
        .normalize("NFKC")
        .trim()
        .toLowerCase()
        .replace(/\s+/g, "-")
        .replace(/[^0-9a-z_\-\u4e00-\u9fff]/g, "")
        .replace(/^[-_]+|[-_]+$/g, "")
        .slice(0, MAX_ACHIEVEMENT_ID_LENGTH);
    return normalize(value) || normalize(fallback) || "achievement";
  }

  function getCustomAchievementStorageId(authorId, fallback = "achievement") {
    return `${CUSTOM_ACHIEVEMENT_PREFIX}${getSafeAchievementAuthorId(authorId, fallback)}`;
  }

  function getLocalizedBlockValue(block, key, fallback, options = {}) {
    if (typeof options.getLocalizedValue === "function") {
      return options.getLocalizedValue(block, key, fallback);
    }
    return block?.[key] ?? fallback;
  }

  function sanitizeAchievementUnlockBlock(block = {}, options = {}) {
    const fallbackId = cleanText(block.id, "achievement", MAX_ACHIEVEMENT_ID_LENGTH);
    const authorId = getSafeAchievementAuthorId(block.achievementId, fallbackId);
    const name = cleanText(
      getLocalizedBlockValue(block, "title", block.title, options),
      "新的成就",
      80
    );
    return {
      id: getCustomAchievementStorageId(authorId),
      authorId,
      name,
      title: name,
      description: cleanText(
        getLocalizedBlockValue(block, "description", block.description, options),
        "完成这段剧情时解锁。",
        240
      ),
      category: cleanText(
        getLocalizedBlockValue(block, "category", block.category, options),
        "剧情里程碑",
        48
      ),
      requirement: cleanText(
        getLocalizedBlockValue(block, "requirement", block.requirement, options),
        "推进到指定剧情",
        120
      ),
      hiddenBeforeUnlock: block.hiddenBeforeUnlock === true,
      iconAssetId: cleanText(block.iconAssetId, "", 128),
      kind: "custom",
      progressTarget: 1,
    };
  }

  function collectCustomAchievementDefinitions(scenes = [], options = {}) {
    const definitions = [];
    const byId = new Map();

    (Array.isArray(scenes) ? scenes : []).forEach((scene) => {
      (Array.isArray(scene?.blocks) ? scene.blocks : []).forEach((block, blockIndex) => {
        if (block?.type !== "achievement_unlock") {
          return;
        }
        const definition = {
          ...sanitizeAchievementUnlockBlock(block, options),
          sceneId: cleanText(scene?.id, "", 128),
          blockId: cleanText(block?.id, "", 128),
          blockIndex,
          duplicateCount: 0,
        };
        const existing = byId.get(definition.id);
        if (existing) {
          existing.duplicateCount += 1;
          return;
        }
        byId.set(definition.id, definition);
        definitions.push(definition);
      });
    });

    return definitions.map((definition) => ({ ...definition }));
  }

  function toIdSet(value) {
    if (value instanceof Set) {
      return new Set(value);
    }
    if (Array.isArray(value)) {
      return new Set(value);
    }
    if (value && typeof value.keys === "function") {
      return new Set(value.keys());
    }
    return new Set();
  }

  function buildAutomaticAchievementDefinitions(metrics = {}, unlockedAchievementIds = []) {
    const unlocked = toIdSet(unlockedAchievementIds);
    const definitions = [];
    const sceneCount = Math.max(0, Number(metrics.sceneCount) || 0);
    const choiceBlockCount = Math.max(0, Number(metrics.choiceBlockCount) || 0);
    const characterCount = Math.max(0, Number(metrics.characterCount) || 0);
    const unlockedCharacterCount = Math.max(0, Number(metrics.unlockedCharacterCount) || 0);
    const galleryCount = Math.max(0, Number(metrics.galleryCount) || 0);
    const unlockedCgCount = Math.max(0, Number(metrics.unlockedCgCount) || 0);
    const musicCount = Math.max(0, Number(metrics.musicCount) || 0);
    const unlockedBgmCount = Math.max(0, Number(metrics.unlockedBgmCount) || 0);
    const endingCount = Math.max(0, Number(metrics.endingCount) || 0);
    const unlockedEndingCount = Math.max(0, Number(metrics.unlockedEndingCount) || 0);
    const add = (definition) => definitions.push({ kind: "automatic", hiddenBeforeUnlock: false, ...definition });

    if (sceneCount > 0) {
      add({
        id: "first_start",
        name: "初次启程",
        category: "剧情里程碑",
        description: "第一次正式开始试玩这个项目。",
        requirement: "点击开始试玩 1 次",
        progressCurrent: unlocked.has("first_start") ? 1 : 0,
        progressTarget: 1,
      });
    }
    if (choiceBlockCount > 0) {
      add({
        id: "first_choice",
        name: "分岔路口",
        category: "剧情里程碑",
        description: "第一次在剧情里做出一个选项分支。",
        requirement: "做出 1 次选项选择",
        progressCurrent: unlocked.has("first_choice") ? 1 : 0,
        progressTarget: 1,
      });
    }
    if (characterCount > 0) {
      add({
        id: "first_character",
        name: "初次相遇",
        category: "人物收集",
        description: "第一次在剧情里见到角色，人物档案馆开始亮起。",
        requirement: "解锁 1 位角色",
        progressCurrent: Math.min(unlockedCharacterCount, 1),
        progressTarget: 1,
      });
      add({
        id: "all_characters",
        name: "全员到齐",
        category: "人物收集",
        description: "把这个项目里的所有角色都收录进图鉴里。",
        requirement: `收录全部 ${characterCount} 位角色`,
        progressCurrent: Math.min(unlockedCharacterCount, characterCount),
        progressTarget: characterCount,
      });
    }
    if (galleryCount > 0) {
      add({
        id: "first_cg",
        name: "回想开幕",
        category: "EXTRA 收集",
        description: "第一次在剧情里解锁一张 CG。",
        requirement: "解锁 1 张 CG",
        progressCurrent: Math.min(unlockedCgCount, 1),
        progressTarget: 1,
      });
      add({
        id: "all_cg",
        name: "回想收藏家",
        category: "EXTRA 收集",
        description: "把这个项目里的所有 CG 都解锁进回想馆。",
        requirement: `解锁全部 ${galleryCount} 张 CG`,
        progressCurrent: Math.min(unlockedCgCount, galleryCount),
        progressTarget: galleryCount,
      });
    }
    if (musicCount > 0) {
      add({
        id: "first_bgm",
        name: "旋律初响",
        category: "EXTRA 收集",
        description: "第一次在剧情里听到并解锁一首 BGM。",
        requirement: "解锁 1 首 BGM",
        progressCurrent: Math.min(unlockedBgmCount, 1),
        progressTarget: 1,
      });
      add({
        id: "all_bgm",
        name: "全曲收藏",
        category: "EXTRA 收集",
        description: "把这个项目里的所有 BGM 都收录进音乐鉴赏。",
        requirement: `解锁全部 ${musicCount} 首 BGM`,
        progressCurrent: Math.min(unlockedBgmCount, musicCount),
        progressTarget: musicCount,
      });
    }
    if (endingCount > 0) {
      add({
        id: "first_ending",
        name: "终幕初见",
        category: "路线回收",
        description: "第一次真正抵达某条路线的结局。",
        requirement: "回收 1 个结局",
        progressCurrent: Math.min(unlockedEndingCount, 1),
        progressTarget: 1,
      });
      add({
        id: "all_endings",
        name: "全结局制霸",
        category: "路线回收",
        description: "把所有可回收结局都点亮进结局回收馆。",
        requirement: `回收全部 ${endingCount} 个结局`,
        progressCurrent: Math.min(unlockedEndingCount, endingCount),
        progressTarget: endingCount,
      });
    }
    return definitions;
  }

  function buildAchievementDefinitions(options = {}) {
    const unlocked = toIdSet(options.unlockedAchievementIds);
    const automatic = buildAutomaticAchievementDefinitions(options.metrics, unlocked);
    const custom = collectCustomAchievementDefinitions(options.scenes, options).map((definition) => ({
      ...definition,
      iconUrl:
        definition.iconAssetId && typeof options.getAssetUrl === "function"
          ? options.getAssetUrl(definition.iconAssetId)
          : "",
      progressCurrent: unlocked.has(definition.id) ? 1 : 0,
      progressTarget: 1,
    }));
    return [...custom, ...automatic];
  }

  function sanitizeAchievementProgressEntries(source, definitions = []) {
    const validIds = new Set((Array.isArray(definitions) ? definitions : []).map((item) => item?.id).filter(Boolean));
    return (source && typeof source === "object" ? Object.entries(source) : []).filter(
      ([achievementId, unlockedAt]) =>
        validIds.has(achievementId) && typeof unlockedAt === "string" && unlockedAt.trim()
    );
  }

  function getAchievementPresentation(definition = {}, unlocked = false) {
    const hidden = definition.hiddenBeforeUnlock === true && !unlocked;
    return {
      name: hidden ? "隐藏成就" : cleanText(definition.name, "未命名成就", 80),
      description: hidden ? "继续探索后才能揭晓这个成就。" : cleanText(definition.description, "尚未填写说明。", 240),
      category: hidden ? "隐藏收集" : cleanText(definition.category, "剧情里程碑", 48),
      requirement: hidden ? "条件尚未公开" : cleanText(definition.requirement, "推进剧情后解锁", 120),
      iconUrl: hidden ? "" : cleanText(definition.iconUrl, "", 1000),
      hidden,
    };
  }

  global.CanvasiaRuntimeAchievements = Object.freeze({
    CUSTOM_ACHIEVEMENT_PREFIX,
    MAX_ACHIEVEMENT_ID_LENGTH,
    getSafeAchievementAuthorId,
    getCustomAchievementStorageId,
    sanitizeAchievementUnlockBlock,
    collectCustomAchievementDefinitions,
    buildAutomaticAchievementDefinitions,
    buildAchievementDefinitions,
    sanitizeAchievementProgressEntries,
    getAchievementPresentation,
  });
})(typeof window !== "undefined" ? window : globalThis);

File: native_runtime/runtime_achievements.py
from __future__ import annotations

import re
import unicodedata
from collections.abc import Callable, Iterable


CUSTOM_ACHIEVEMENT_PREFIX = "custom:"
MAX_ACHIEVEMENT_ID_LENGTH = 64
_UNSAFE_ACHIEVEMENT_ID_PATTERN = re.compile(r"[^0-9a-z_\-\u4e00-\u9fff]+")


def _clean_text(value: object, fallback: object = "", max_length: int = 240) -> str:
    text = str(value or "").strip() or str(fallback or "").strip()
    return text[:max_length]


def get_safe_achievement_author_id(value: object, fallback: object = "achievement") -> str:
    def normalize(candidate: object) -> str:
        text = unicodedata.normalize("NFKC", str(candidate or "")).strip().lower()
        text = re.sub(r"\s+", "-", text)
        text = _UNSAFE_ACHIEVEMENT_ID_PATTERN.sub("", text)
        return text.strip("-_")[:MAX_ACHIEVEMENT_ID_LENGTH]

    return normalize(value) or normalize(fallback) or "achievement"


def get_custom_achievement_storage_id(author_id: object, fallback: object = "achievement") -> str:
    return f"{CUSTOM_ACHIEVEMENT_PREFIX}{get_safe_achievement_author_id(author_id, fallback)}"


def _localized_value(
    block: dict,
    key: str,
    fallback: object,
    localize_value: Callable[[dict, str, object], object] | None,
) -> object:
    if callable(localize_value):
        return localize_value(block, key, fallback)
    return block.get(key, fallback)


def sanitize_achievement_unlock_block(
    block: dict | None,
    *,
    localize_value: Callable[[dict, str, object], object] | None = None,
) -> dict:
    source = block if isinstance(block, dict) else {}
    fallback_id = _clean_text(source.get("id"), "achievement", MAX_ACHIEVEMENT_ID_LENGTH)
    author_id = get_safe_achievement_author_id(source.get("achievementId"), fallback_id)
    name = _clean_text(
        _localized_value(source, "title", source.get("title"), localize_value),
        "新的成就",
        80,
    )
    return {
        "id": get_custom_achievement_storage_id(author_id),
        "authorId": author_id,
        "name": name,
        "title": name,
        "description": _clean_text(
            _localized_value(source, "description", source.get("description"), localize_value),
            "完成这段剧情时解锁。",
            240,
        ),
        "category": _clean_text(
            _localized_value(source, "category", source.get("category"), localize_value),
            "剧情里程碑",
            48,
        ),
        "requirement": _clean_text(
            _localized_value(source, "requirement", source.get("requirement"), localize_value),
            "推进到指定剧情",
            120,
        ),
        "hiddenBeforeUnlock": source.get("hiddenBeforeUnlock") is True,
        "iconAssetId": _clean_text(source.get("iconAssetId"), "", 128),
        "kind": "custom",
        "progressTarget": 1,
    }


def collect_custom_achievement_definitions(
    chapters: list[dict] | None,
    *,
    localize_value: Callable[[dict, str, object], object] | None = None,
) -> list[dict]:
    definitions: list[dict] = []
    by_id: dict[str, dict] = {}
    for chapter in chapters if isinstance(chapters, list) else []:
        for scene in chapter.get("scenes") or []:
            for block_index, block in enumerate(scene.get("blocks") or []):
                if not isinstance(block, dict) or block.get("type") != "achievement_unlock":
                    continue
                definition = {
                    **sanitize_achievement_unlock_block(block, localize_value=localize_value),
                    "sceneId": _clean_text(scene.get("id"), "", 128),
                    "blockId": _clean_text(block.get("id"), "", 128),
                    "blockIndex": block_index,
                    "duplicateCount": 0,
                }
                existing = by_id.get(definition["id"])
                if existing:
                    existing["duplicateCount"] += 1
                    continue
                by_id[definition["id"]] = definition
                definitions.append(definition)
    return [dict(definition) for definition in definitions]


def build_native_automatic_achievement_entries(metrics: dict | None = None) -> list[dict]:
    source = metrics if isinstance(metrics, dict) else {}
    character_count = max(0, int(source.get("characterCount") or 0))
    unlocked_characters = max(0, int(source.get("unlockedCharacterCount") or 0))
    gallery_count = max(0, int(source.get("galleryCount") or 0))
    unlocked_cg = max(0, int(source.get("unlockedCgCount") or 0))
    music_count = max(0, int(source.get("musicCount") or 0))
    unlocked_bgm = max(0, int(source.get("unlockedBgmCount") or 0))
    ending_count = max(0, int(source.get("endingCount") or 0))
    unlocked_endings = max(0, int(source.get("unlockedEndingCount") or 0))
    entries = [
        {
            "id": "first_start",
            "name": "初次启动",
            "subtitle": "第一次进入原生 Runtime",
            "notes": "启动一次原生 Runtime 即可点亮。",
            "unlocked": bool(source.get("hasScenes")),
        }
    ]
    collections = (
        ("character", "characters", "角色", character_count, unlocked_characters, "收录"),
        ("cg", "cg", "CG", gallery_count, unlocked_cg, "回收"),
        ("bgm", "bgm", "乐曲", music_count, unlocked_bgm, "解锁"),
        ("ending", "endings", "结局", ending_count, unlocked_endings, "回收"),
    )
    for key, collection_id, label, total, current, verb in collections:
        if total <= 0:
            continue
        entries.extend(
            [
                {
                    "id": f"first_{key}",
                    "name": f"{label}初见",
                    "subtitle": f"已{verb}{label} {current} / {total}",
                    "notes": f"任意{label}完成{verb}后点亮。",
                    "unlocked": current > 0,
                },
                {
                    "id": f"all_{collection_id}",
                    "name": f"{label}全{verb}",
                    "subtitle": f"已{verb}{label} {current} / {total}",
                    "notes": f"{verb}全部{label}后点亮。",
                    "unlocked": current >= total,
                },
            ]
        )
    return entries


def build_native_achievement_archive_entries(
    chapters: list[dict] | None,
    *,
    metrics: dict | None = None,
    unlocked_custom_ids: Iterable[str] | None = None,
    localize_value: Callable[[dict, str, object], object] | None = None,
) -> list[dict]:
    unlocked_ids = {str(value) for value in unlocked_custom_ids or [] if str(value).strip()}
    entries: list[dict] = []
    for definition in collect_custom_achievement_definitions(chapters, localize_value=localize_value):
        unlocked = definition["id"] in unlocked_ids
        hidden = definition.get("hiddenBeforeUnlock") is True and not unlocked
        name = "隐藏成就" if hidden else definition["name"]
        description = "继续探索后才能揭晓这个成就。" if hidden else definition["description"]
        category = "隐藏收集" if hidden else definition["category"]
        requirement = "条件尚未公开" if hidden else definition["requirement"]
        entries.append(
            {
                **definition,
                "name": name,
                "subtitle": f"{category} · {'已达成' if unlocked else '未达成'}",
                "notes": description if unlocked else f"目标：{requirement}",
                "actionLabel": "查看成就",
                "actionEnabled": unlocked,
                "previewAssetId": "" if hidden else definition["iconAssetId"],
                "previewText": description,
                "unlocked": unlocked,
                "hidden": hidden,
            }
        )

    entries.extend(
        {
            **entry,
            "kind": "automatic",
            "actionLabel": "查看成就",
            "actionEnabled": bool(entry.get("unlocked")),
            "previewText": entry["notes"],
        }
        for entry in build_native_automatic_achievement_entries(metrics)
    )
    return entries
